compiler/rustc_mir_transform/src/check_pointers.rs

```diff
@@ -0,0 +1,161 @@
+use rustc_hir::lang_items::LangItem;
+use rustc_index::IndexVec;
+use rustc_middle::mir::visit::{MutatingUseContext, NonMutatingUseContext, PlaceContext, Visitor};
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use tracing::{debug, trace};
+
+pub(crate) fn check_pointers<'a, 'tcx, F>(
```

Comment:
Please add an extensive doc comment explaining what this does. I currently find this logic too hard to follow.

Reply:
FYI, this is a cut-and-paste of the previous implementation. If you think the use of PlaceMention here is confusing, we could indeed revert that. But I am not sure such concerns belong in this PR.

Reply:
Previously this was an internal implementation detail of one pass. Now it's a shared utility. A shared utility should document its contract with its users.

Reply:
Sorry, that's a very fair point. I refactored the interface to be explained better and be more intuitive, PTAL.